### PR TITLE
Feature: Edges have a "bundled" getter for when they represents a bundled dependency

### PR DESCRIPTION
**Pull request by kumavis**
when analyzing a dep graph its useful to see if an Edge is a `bundledDependencies` edge. this is especially useful when trying to understand the provenance of a package's content.

This PR adds a `bundled` getter property to the `Edge.prototype` which returns true if its a `bundledDependencies` edge or not.

## References
  Related to https://github.com/npm/cli/pull/2750

### CODE REVIEW

**Comment by isaacs** [comment]
Since it's only checking that the edge.from is the bundler (and not that the node is in a bundle at all), it would be more efficient to check `edge.from.package.bundleDependencies.includes(edge.name)`. This would also mark the edge as a bundling edge, even if it is not resolved.

**Comment by isaacs** [comment]
This _would_ be really handy to make the code in a few places easier to follow, I like the idea a lot.

**Comment by kumavis** [comment]
yeah sounds good, I was thrown off by this part and wasn't sure what the appropriate check was
https://github.com/npm/arborist/blob/6f99a5e7db8fe21ace364bf7533fe5447149cc25/lib/node.js#L435-L440

any situation where it would be bundled but `this.to.name` would differ from whats in `this.from.package.bundleDependencies`?

File: lib/edge.js [hunk 1]
```diff
@@ -92,11 +92,17 @@ class Edge {
...
+      bundled: this.bundled,
       ...(error ? { error } : {}),
       ...(from ? { from: from.explain(null, seen) } : {}),
     }
   }
 
+  get bundled () {
+    const fromBundleDependencies = this.from && this.from.package && this.from.package.bundleDependencies
+    return (fromBundleDependencies && this.to) ? fromBundleDependencies.includes(this.to.name) : false
+  }
```
Comment (ljharb) [inline lib/edge.js:104]: is this worth memoizing?

Comment (kumavis) [reply lib/edge.js:104]: its currently only used in `explain`

Comment (isaacs) [reply lib/edge.js:104]: `this.from` will always be set unless the edge is in a detached state.

`this.from.package` will always be set (it'll at least be `{}`).

`this.name` is the name of the edge, which will typically always match the name of the node it resolves to.  (A node name doesn't _always_ match its location in `node_modules`, but only because a node might not _be_ in `node_modules`, but in that case, an edge can't resolve to it.)

So, this bit can be simplified to:

```js
get bundled () {
  if (this.error === 'DETACHED')
    return false
  const { bundleDependencies = [] } = this.from
  return bundleDependencies.includes(this.name)
}
```

File: lib/edge.js [hunk 2]
```diff
@@ -92,11 +92,17 @@ class Edge {
       type: this.type,
       name: this.name,
       spec: this.spec,
+      bundled: this.bundled,
```
Comment (isaacs) [inline lib/edge.js:95]: This gets quite noisy.

I recommend changing the line above to `const { error, from, bundled } = this`, and then change this line to

```js
  ...(bundled ? { bundled } : {}),
```

so it's filtered out like `from` and `error` when not relevant.

File: test/edge.js [hunk 3]
```diff
@@ -440,3 +458,66 @@ t.match(
...
+  edgesOut: new Map(),
+  edgesIn: new Set(),
+  explain: () => 'bundleChild explanation',
+  package: { name: 'bundle-child', version: '1.2.3' },
+  get version () {
+    return this.package.version
+  },
+  isTop: false,
+  parent: top,
+  getBundler () {
```
Comment (isaacs) [inline test/edge.js:473]: I don't think we need `getBundler` in the test if the Edge.bundled getter isn't using it, yes?

File: lib/edge.js (hunk 2 above)
Comment (ljharb) [reply lib/edge.js:95]: or
```
...(bundled && { bundled }),
```

File: lib/edge.js (hunk 1 above)
Comment (kumavis) [reply lib/edge.js:104]: my extra defensive code here is partially dealing with the tests which are incomplete non-normalized pojos

File: lib/edge.js (hunk 2 above)
Comment (kumavis) [reply lib/edge.js:95]: i feel more sympathy for v8 wanting consistent object shape than noise in the snapshots, but can do

File: lib/edge.js (hunk 1 above)
Comment (kumavis) [reply lib/edge.js:104]: > this.name is the name of the edge, which will typically always match the name of the node it resolves to. 

when would it not match the name of the node

Comment (kumavis) [reply lib/edge.js:104]: is it better to check if its in a detached state which has the side effect of not having a `from`, rather than just checking directly for a from?

File: lib/edge.js (hunk 2 above)
Comment (kumavis) [reply lib/edge.js]: did this

File: test/edge.js (hunk 3 above)
Comment (kumavis) [reply test/edge.js]: done

File: lib/edge.js (hunk 1 above)
Comment (kumavis) [reply lib/edge.js]: did this, but checked directly for `this.from` instead of the error state

**Comment by kumavis** [comment]
neat. im happy with this
